File: clients/ios/AwareChat-iOS/AwareChat-iOS/Core/Services/MessagingService.swift
import Foundation

/// App-owned synchronization. Only its event loop advances the connection and FIFO outbox.
actor MessagingService: MessagingServiceProtocol {
  private let socket: any WebSocketClientProtocol
  private let users: any UserLocalRepository
  private let messages: any MessageLocalRepository
  private let clock: any MessagingClock

  private var state = MessagingConnectionState.disconnected
  private var observers: [UUID: AsyncStream<MessagingConnectionState>.Continuation] = [:]
  private var issueObservers: [UUID: AsyncStream<MessagingIssue>.Continuation] = [:]
  private var worker: Task<Void, Never>?
  private var runID: UUID?
  private var inbox: AsyncStream<MessagingSignal>.Continuation?
  private var sessionID: UUID?
  private var identity: UserDTO?
  private var reader: Task<Void, Never>?
  private var writes: [UUID: Task<Void, Never>] = [:]
  private var deadline: Task<Void, Never>?
  private var deadlineID: UUID?
  private var retryTask: Task<Void, Never>?
  private var retryID: UUID?
  private var retryPolicy = MessagingRetryPolicy()
  private var inFlight: UUID?
  private var retryMessageID: UUID?
  private var attemptedMessages: Set<UUID> = []

  init(
    socket: any WebSocketClientProtocol,
    users: any UserLocalRepository,
    messages: any MessageLocalRepository,
    clock: any MessagingClock = SystemMessagingClock()
  ) {
    self.socket = socket
    self.users = users
    self.messages = messages
    self.clock = clock
  }

  /// Save/reuse the identity through UserLocalRepository before starting.
  /// Repeated starts are idempotent. Await stop() before an explicit restart.
  func start() {
    guard worker == nil else { return }
    let id = UUID()
    let (stream, continuation) = AsyncStream<MessagingSignal>.makeStream()
    runID = id
    inbox = continuation
    retryPolicy.reset()
    retryMessageID = nil
    attemptedMessages.removeAll()
    setState(.connecting)
    worker = Task { [weak self] in await self?.run(stream: stream, id: id) }
    continuation.yield(.connect)
  }

  func stop() async {
    let id = runID
    let activeWorker = worker
    activeWorker?.cancel()
    inbox?.finish()
    await activeWorker?.value
    // A newer explicit start must not be erased by this stop's suspended continuation.
    if runID == id || runID == nil { setState(.disconnected) }
  }

  @discardableResult
  func sendMessage(text: String, receiverId: UUID) async throws -> LocalMessage {
    let message = try await messages.enqueue(text: text, receiverId: receiverId, createdAt: clock.now())
    inbox?.yield(.wakeOutbox)
    return message
  }

  func connectionState() -> MessagingConnectionState { state }

  func observeConnectionState() -> AsyncStream<MessagingConnectionState> {
    let id = UUID()
    let (stream, continuation) = AsyncStream<MessagingConnectionState>.makeStream(
      bufferingPolicy: .bufferingNewest(1)
    )
    observers[id] = continuation
    continuation.yield(state)
    continuation.onTermination = { [weak self] _ in
      Task { await self?.removeObserver(id) }
    }
    return stream
  }

  /// Subscribe before requesting work. Issues are transient, not another copy of message history.
  func observeIssues() -> AsyncStream<MessagingIssue> {
    let id = UUID()
    let (stream, continuation) = AsyncStream<MessagingIssue>.makeStream(bufferingPolicy: .bufferingNewest(1))
    issueObservers[id] = continuation
    continuation.onTermination = { [weak self] _ in
      Task { await self?.removeIssueObserver(id) }
    }
    return stream
  }

  private func run(stream: AsyncStream<MessagingSignal>, id: UUID) async {
    do {
      try await messages.recoverInterruptedSends()
      for await signal in stream {
        try Task.checkCancellation()
        try await handle(signal)
      }
    } catch {
      if !Task.isCancelled { setState(.connectionFailure(failure(from: error))) }
    }
    await closeSession()
    do { try await messages.recoverInterruptedSends() }
    catch { setState(.connectionFailure(failure(from: error))) }
    if runID == id {
      inbox?.finish()
      inbox = nil
      worker = nil
      runID = nil
    }
  }

  private func handle(_ signal: MessagingSignal) async throws {
    switch signal {
    case .connect:
      try await connect()
    case .wakeOutbox:
      try await flushNext()
    case .event(let session, let event):
      guard sessionID == session else { return }
      try await handle(event)
    case .lost(let session, let error):
      guard sessionID == session else { return }
      if case .webSocketClosed(code: 4001, reason: _) = error {
        throw MessagingFailure.sessionReplaced
      }
      try await reconnect(after: .network(error))
    case .writeFinished(let session, let writeID, let error):
      guard sessionID == session else { return }
      writes.removeValue(forKey: writeID)
      if let error { try await reconnect(after: .network(error)) }
    case .deadline(let session, let token):
      guard sessionID == session, deadlineID == token else { return }
      try await reconnect(after: identity == nil ? .identificationTimedOut : .acceptanceTimedOut)
    case .retry(let token):
      guard retryID == token else { return }
      retryTask = nil
      retryID = nil
      if sessionID == nil { try await connect() }
      else { try await flushNext() }
    }
  }

  private func connect() async throws {
    guard let current = try await users.currentUser() else { throw PersistenceError.missingIdentity }
    let user = try current.wireIdentity()
    try Task.checkCancellation()
    setState(.connecting)
    let session = UUID()
    sessionID = session
    let events = await socket.connect()
    try Task.checkCancellation()
    guard let inbox else { return }
    reader = Task {
      do {
        for try await event in events {
          try Task.checkCancellation()
          inbox.yield(.event(session: session, event))
        }
        if !Task.isCancelled { inbox.yield(.lost(session: session, .notConnected)) }
      } catch {
        if !Task.isCancelled { inbox.yield(.lost(session: session, NetworkError.wrapping(error))) }
      }
    }
    setDeadline(MessagingRetryPolicy.identificationTimeout)
    write(.identify(user: user))
  }

  private func handle(_ event: ServerEvent) async throws {
    switch event {
    case .identityAccepted(let accepted):
      guard identity == nil else { return }
      guard let current = try await users.currentUser(), try current.wireIdentity() == accepted else {
        throw MessagingFailure.unexpectedIdentity
      }
      try Task.checkCancellation()
      try await users.completeRegistration(acceptedUser: accepted)
      try Task.checkCancellation()
      identity = accepted
      cancelDeadline()
      if retryMessageID == nil { retryPolicy.reset() }
      setState(.connected)
      try await flushNext()
    case .syncCompleted:
      // Replay completion is not a persistence ACK or an identification gate.
      break
    case .incomingMessage(let message):
      guard identity != nil else { throw MessagingFailure.network(.invalidWebSocketFrame) }
      try await messages.persistIncoming(message, receivedAt: clock.now())
      try Task.checkCancellation()
      write(.messagePersisted(messageId: message.messageId))
    case .messageAccepted(let id, let timestamp):
      guard identity != nil, attemptedMessages.contains(id),
            let message = try await messages.message(id: id), message.direction == .outgoing else { return }
      try await messages.markAccepted(id: id, serverReceivedAt: timestamp)
      try Task.checkCancellation()
      attemptedMessages.remove(id)
      if inFlight == id || retryMessageID == id {
        inFlight = nil
        retryMessageID = nil
        cancelDeadline()
        cancelRetry()
        retryPolicy.reset()
        try await flushNext()
      }
    case .protocolError(let event):
      try await handleProtocolError(event)
    }
  }

  private func handleProtocolError(_ event: ProtocolErrorEvent) async throws {
    let error = event.error
    let issue = MessagingIssue(messageId: event.messageId.flatMap(UUID.init(uuidString:)), error: error)
    for observer in issueObservers.values { observer.yield(issue) }
    if error.code == "SESSION_REPLACED" { throw MessagingFailure.sessionReplaced }
    // These errors describe recipient acknowledgements, not outgoing message acceptance.
    if error.code == "UNKNOWN_MESSAGE" || error.code == "NOT_RECEIVER" { return }
    if let rawID = event.messageId, let id = UUID(uuidString: rawID) {
      guard id == inFlight || id == retryMessageID,
            let message = try await messages.message(id: id),
            message.direction == .outgoing,
            message.state == .sending || message.state == .pendingToSend else { return }
      try await messages.markRejected(id: id, retryable: error.isRetryable)
      try Task.checkCancellation()
      inFlight = nil
      cancelDeadline()
      if error.isRetryable {
        retryMessageID = id
        scheduleRetry()
      } else {
        attemptedMessages.remove(id)
        retryMessageID = nil
        cancelRetry()
        retryPolicy.reset()
        try await flushNext()
      }
    } else if event.messageId == nil {
      if error.isRetryable { try await reconnect(after: .server(error)) }
      else { throw MessagingFailure.server(error) }
    }
  }

  private func flushNext() async throws {
    guard identity != nil, inFlight == nil, retryID == nil else { return }
    guard let message = try await messages.pendingMessages().first else { return }
    try Task.checkCancellation()
    try await messages.markSending(id: message.id)
    try Task.checkCancellation()
    inFlight = message.id
    attemptedMessages.insert(message.id)
    setDeadline(MessagingRetryPolicy.acceptanceTimeout)
    write(.sendMessage(message: try message.outgoingPayload()))
  }

  private func reconnect(after failure: MessagingFailure) async throws {
    if let inFlight { retryMessageID = inFlight }
    await closeSession()
    try Task.checkCancellation()
    try await messages.recoverInterruptedSends()
    try Task.checkCancellation()
    setState(.connectionFailure(failure))
    scheduleRetry()
  }

  private func write(_ event: ClientEvent) {
    guard let session = sessionID, let inbox else { return }
    let id = UUID()
    let socket = socket
    writes[id] = Task {
      do {
        try Task.checkCancellation()
        try await socket.send(event)
        if !Task.isCancelled { inbox.yield(.writeFinished(session: session, id, nil)) }
      } catch {
        if !Task.isCancelled {
          inbox.yield(.writeFinished(session: session, id, NetworkError.wrapping(error)))
        }
      }
    }
  }

  private func setDeadline(_ duration: Duration) {
    cancelDeadline()
    guard let session = sessionID, let inbox else { return }
    let token = UUID()
    deadlineID = token
    let clock = clock
    deadline = Task {
      do {
        try await clock.sleep(for: duration)
        try Task.checkCancellation()
        inbox.yield(.deadline(session: session, token))
      } catch { /* Cancellation invalidates this timer. */ }
    }
  }

  private func scheduleRetry() {
    cancelRetry()
    guard let inbox else { return }
    let token = UUID()
    retryID = token
    let delay = retryPolicy.nextDelay()
    let clock = clock
    retryTask = Task {
      do {
        try await clock.sleep(for: delay)
        try Task.checkCancellation()
        inbox.yield(.retry(token))
      } catch { /* Cancellation invalidates this timer. */ }
    }
  }

  private func closeSession() async {
    sessionID = nil
    identity = nil
    inFlight = nil
    cancelDeadline()
    cancelRetry()
    reader?.cancel()
    reader = nil
    for task in writes.values { task.cancel() }
    writes.removeAll()
    await socket.disconnect()
  }

  private func cancelDeadline() {
    deadline?.cancel()
    deadline = nil
    deadlineID = nil
  }

  private func cancelRetry() {
    retryTask?.cancel()
    retryTask = nil
    retryID = nil
  }

  private func setState(_ newState: MessagingConnectionState) {
    guard state != newState else { return }
    state = newState
    for observer in observers.values { observer.yield(newState) }
  }

  private func removeObserver(_ id: UUID) { observers.removeValue(forKey: id) }
  private func removeIssueObserver(_ id: UUID) { issueObservers.removeValue(forKey: id) }

  private func failure(from error: any Error) -> MessagingFailure {
    if let error = error as? MessagingFailure { return error }
    if let error = error as? PersistenceError { return .storage(error) }
    return .network(NetworkError.wrapping(error))
  }
}

private nonisolated enum MessagingSignal: Sendable {
  case connect
  case wakeOutbox
  case event(session: UUID, ServerEvent)
  case lost(session: UUID, NetworkError)
  case writeFinished(session: UUID, UUID, NetworkError?)
  case deadline(session: UUID, UUID)
  case retry(UUID)
}

File: clients/android/AwareChat-Android/app/src/main/java/com/example/awarechat_android/core/service/MessagingService.kt
package com.example.awarechat_android.core.service

import com.example.awarechat_android.core.network.NetworkException
import com.example.awarechat_android.core.network.WebSocketClientContract
import com.example.awarechat_android.core.network.WebSocketConnectionState
import com.example.awarechat_android.core.persistence.database.PersistenceException
import com.example.awarechat_android.core.persistence.messages.MessageDirection
import com.example.awarechat_android.core.persistence.messages.MessageLocalRepository
import com.example.awarechat_android.core.persistence.messages.MessageState
import com.example.awarechat_android.core.persistence.users.UserLocalRepository
import com.example.awarechat_android.core.protocol.ClientEvent
import com.example.awarechat_android.core.protocol.ProtocolErrorEvent
import com.example.awarechat_android.core.protocol.ServerEvent
import com.example.awarechat_android.core.protocol.UserDto
import java.util.UUID
import kotlin.coroutines.cancellation.CancellationException
import kotlin.time.Duration
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancelAndJoin
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/** App-owned synchronization. One worker serializes connection and outbox state. */
class MessagingService(
    private val socket: WebSocketClientContract,
    private val users: UserLocalRepository,
    private val messages: MessageLocalRepository,
    private val clock: MessagingClock = SystemMessagingClock(),
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO),
) : MessagingServiceContract {
    private val lock = Any()
    private val mutableConnectionState =
        MutableStateFlow<MessagingConnectionState>(MessagingConnectionState.Disconnected)
    private val mutableIssues = MutableSharedFlow<MessagingIssue>(extraBufferCapacity = 1)
    private var runId: UUID? = null
    private var worker: Job? = null
    private var inbox: Channel<Signal>? = null

    override val connectionState: StateFlow<MessagingConnectionState> =
        mutableConnectionState.asStateFlow()
    override val issues: SharedFlow<MessagingIssue> = mutableIssues.asSharedFlow()

    override fun start() {
        synchronized(lock) {
            if (worker?.isActive == true) return

            val id = UUID.randomUUID()
            val signals = Channel<Signal>(Channel.UNLIMITED)
            runId = id
            inbox = signals
            publishState(id, MessagingConnectionState.Connecting)
            worker = scope.launch {
                Worker(
                    inbox = signals,
                    socket = socket,
                    users = users,
                    messages = messages,
                    clock = clock,
                    publishState = { publishState(id, it) },
                    publishIssue = mutableIssues::tryEmit,
                ).run()
                synchronized(lock) {
                    if (runId == id) {
                        worker = null
                        inbox = null
                    }
                }
            }
            signals.trySend(Signal.Connect)
        }
    }

    override suspend fun stop() {
        val active = synchronized(lock) {
            val active = worker
            runId = null
            worker = null
            inbox?.close()
            inbox = null
            mutableConnectionState.value = MessagingConnectionState.Disconnected
            active
        }
        active?.cancelAndJoin()
        socket.disconnect()
    }

    override suspend fun sendMessage(text: String, receiverId: UUID) =
        messages.enqueue(text = text, receiverId = receiverId, createdAt = clock.now()).also {
            synchronized(lock) { inbox }?.trySend(Signal.WakeOutbox)
        }

    private fun publishState(id: UUID, state: MessagingConnectionState) {
        synchronized(lock) {
            if (runId == id) mutableConnectionState.value = state
        }
    }
}

private class Worker(
    private val inbox: Channel<Signal>,
    private val socket: WebSocketClientContract,
    private val users: UserLocalRepository,
    private val messages: MessageLocalRepository,
    private val clock: MessagingClock,
    private val publishState: (MessagingConnectionState) -> Unit,
    private val publishIssue: (MessagingIssue) -> Boolean,
) {
    private lateinit var workerScope: CoroutineScope
    private var sessionId: UUID? = null
    private var identity: UserDto? = null
    private var reader: Job? = null
    private var opening: Job? = null
    private var deadline: Job? = null
    private var deadlineId: UUID? = null
    private var retry: Job? = null
    private var retryId: UUID? = null
    private var retryPolicy = MessagingRetryPolicy()
    private var inFlight: UUID? = null
    private var retryMessageId: UUID? = null
    private val attemptedMessages = mutableSetOf<UUID>()

    suspend fun run() = kotlinx.coroutines.coroutineScope {
        workerScope = this
        try {
            messages.recoverInterruptedSends()
            for (signal in inbox) handle(signal)
        } catch (error: CancellationException) {
            throw error
        } catch (error: Throwable) {
            publishState(MessagingConnectionState.ConnectionFailure(error.toMessagingFailure()))
        } finally {
            withContext(NonCancellable) {
                closeConnection()
                try {
                    messages.recoverInterruptedSends()
                } catch (error: Throwable) {
                    publishState(
                        MessagingConnectionState.ConnectionFailure(error.toMessagingFailure()),
                    )
                }
            }
        }
    }

    private suspend fun handle(signal: Signal) {
        when (signal) {
            Signal.Connect -> connect()
            Signal.WakeOutbox -> flushNext()
            is Signal.Opened -> if (sessionId == signal.sessionId) identify()
            is Signal.Event -> if (sessionId == signal.sessionId) handle(signal.event)
            is Signal.Lost -> if (sessionId == signal.sessionId) handleConnectionLoss(signal.error)
            is Signal.Deadline -> if (
                sessionId == signal.sessionId && deadlineId == signal.deadlineId
            ) {
                reconnect(
                    if (identity == null) {
                        MessagingFailure.IdentificationTimedOut
                    } else {
                        MessagingFailure.AcceptanceTimedOut
                    },
                )
            }
            is Signal.Retry -> if (retryId == signal.retryId) {
                retry = null
                retryId = null
                if (sessionId == null) connect() else flushNext()
            }
        }
    }

    private suspend fun connect() {
        val current = users.currentUser()
            ?: throw PersistenceException(
                com.example.awarechat_android.core.persistence.database.PersistenceFailure.MISSING_IDENTITY,
            )
        val user = current.wireIdentity()
        publishState(MessagingConnectionState.Connecting)
        socket.disconnect()
        val session = UUID.randomUUID()
        sessionId = session
        setDeadline(session, MessagingRetryPolicy.identificationTimeout)

        reader = workerScope.launch {
            try {
                socket.connect().collect { inbox.send(Signal.Event(session, it)) }
                if (sessionId == session) {
                    inbox.send(Signal.Lost(session, NetworkException.NotConnected()))
                }
            } catch (error: CancellationException) {
                throw error
            } catch (error: Throwable) {
                inbox.send(Signal.Lost(session, NetworkException.wrap(error)))
            }
        }
        opening = workerScope.launch {
            val state = socket.connectionState.first {
                it == WebSocketConnectionState.Connected ||
                    it == WebSocketConnectionState.ConnectionFailure
            }
            if (state == WebSocketConnectionState.Connected) {
                inbox.send(Signal.Opened(session))
            }
        }
        pendingIdentity = user
    }

    private var pendingIdentity: UserDto? = null

    private suspend fun identify() {
        val user = pendingIdentity ?: return
        sendOrReconnect(ClientEvent.Identify(user))
    }

    private suspend fun handle(event: ServerEvent) {
        when (event) {
            is ServerEvent.IdentityAccepted -> handleIdentityAccepted(event.user)
            is ServerEvent.SyncCompleted -> Unit
            is ServerEvent.IncomingMessage -> handleIncoming(event)
            is ServerEvent.MessageAccepted -> handleMessageAccepted(event)
            is ServerEvent.ProtocolError -> handleProtocolError(event.event)
        }
    }

    private suspend fun handleIdentityAccepted(accepted: UserDto) {
        if (identity != null) return
        val current = users.currentUser()
        if (current?.wireIdentity() != accepted) throw MessagingFailureException(
            MessagingFailure.UnexpectedIdentity,
        )
        users.completeRegistration(accepted)
        identity = accepted
        pendingIdentity = null
        cancelDeadline()
        if (retryMessageId == null) retryPolicy.reset()
        publishState(MessagingConnectionState.Connected)
        flushNext()
    }

    private suspend fun handleIncoming(event: ServerEvent.IncomingMessage) {
        if (identity == null) throw MessagingFailureException(
            MessagingFailure.Network(NetworkException.InvalidWebSocketFrame()),
        )
        messages.persistIncoming(event.message, clock.now())
        sendOrReconnect(ClientEvent.MessagePersisted(event.message.messageId))
    }

    private suspend fun handleMessageAccepted(event: ServerEvent.MessageAccepted) {
        if (identity == null || event.messageId !in attemptedMessages) return
        val message = messages.message(event.messageId) ?: return
        if (message.direction != MessageDirection.OUTGOING) return

        messages.markAccepted(event.messageId, event.serverReceivedAt)
        attemptedMessages.remove(event.messageId)
        if (inFlight == event.messageId || retryMessageId == event.messageId) {
            inFlight = null
            retryMessageId = null
            cancelDeadline()
            cancelRetry()
            retryPolicy.reset()
            flushNext()
        }
    }

    private suspend fun handleProtocolError(event: ProtocolErrorEvent) {
        val messageId = event.messageId?.let { raw ->
            runCatching { UUID.fromString(raw) }.getOrNull()
        }
        publishIssue(MessagingIssue(messageId, event.error))
        if (event.error.code == "SESSION_REPLACED") {
            throw MessagingFailureException(MessagingFailure.SessionReplaced)
        }
        if (event.error.code == "UNKNOWN_MESSAGE" || event.error.code == "NOT_RECEIVER") return

        if (messageId != null) {
            if (messageId != inFlight && messageId != retryMessageId) return
            val message = messages.message(messageId) ?: return
            if (message.direction != MessageDirection.OUTGOING ||
                (message.state != MessageState.SENDING &&
                    message.state != MessageState.PENDING_TO_SEND)
            ) {
                return
            }
            messages.markRejected(messageId, event.error.isRetryable)
            inFlight = null
            cancelDeadline()
            if (event.error.isRetryable) {
                retryMessageId = messageId
                scheduleRetry()
            } else {
                attemptedMessages.remove(messageId)
                retryMessageId = null
                cancelRetry()
                retryPolicy.reset()
                flushNext()
            }
        } else if (event.error.isRetryable) {
            reconnect(MessagingFailure.Server(event.error))
        } else {
            throw MessagingFailureException(MessagingFailure.Server(event.error))
        }
    }

    private suspend fun flushNext() {
        if (identity == null || inFlight != null || retryId != null) return
        val message = messages.pendingMessages().firstOrNull() ?: return
        messages.markSending(message.message.messageId)
        inFlight = message.message.messageId
        attemptedMessages += message.message.messageId
        setDeadline(requireNotNull(sessionId), MessagingRetryPolicy.acceptanceTimeout)
        sendOrReconnect(ClientEvent.SendMessage(message.outgoingPayload()))
    }

    private suspend fun sendOrReconnect(event: ClientEvent) {
        try {
            socket.send(event)
        } catch (error: CancellationException) {
            throw error
        } catch (error: Throwable) {
            reconnect(MessagingFailure.Network(NetworkException.wrap(error)))
        }
    }

    private suspend fun handleConnectionLoss(error: NetworkException) {
        if (error is NetworkException.WebSocketClosed && error.code == SESSION_REPLACED_CODE) {
            throw MessagingFailureException(MessagingFailure.SessionReplaced)
        }
        reconnect(MessagingFailure.Network(error))
    }

    private suspend fun reconnect(failure: MessagingFailure) {
        inFlight?.let { retryMessageId = it }
        closeConnection()
        messages.recoverInterruptedSends()
        publishState(MessagingConnectionState.ConnectionFailure(failure))
        scheduleRetry()
    }

    private fun setDeadline(session: UUID, duration: Duration) {
        cancelDeadline()
        val token = UUID.randomUUID()
        deadlineId = token
        deadline = workerScope.launch {
            try {
                clock.sleep(duration)
                inbox.send(Signal.Deadline(session, token))
            } catch (_: CancellationException) {
                // Cancellation invalidates this deadline.
            }
        }
    }

    private fun scheduleRetry() {
        cancelRetry()
        val token = UUID.randomUUID()
        retryId = token
        val duration = retryPolicy.nextDelay()
        retry = workerScope.launch {
            try {
                clock.sleep(duration)
                inbox.send(Signal.Retry(token))
            } catch (_: CancellationException) {
                // Cancellation invalidates this retry.
            }
        }
    }

    private suspend fun closeConnection() {
        sessionId = null
        identity = null
        pendingIdentity = null
        inFlight = null
        cancelDeadline()
        opening?.cancelAndJoin()
        opening = null
        reader?.cancelAndJoin()
        reader = null
        socket.disconnect()
    }

    private fun cancelDeadline() {
        deadline?.cancel()
        deadline = null
        deadlineId = null
    }

    private fun cancelRetry() {
        retry?.cancel()
        retry = null
        retryId = null
    }

    private companion object {
        const val SESSION_REPLACED_CODE = 4001
    }
}

private sealed interface Signal {
    data object Connect : Signal
    data object WakeOutbox : Signal
    data class Opened(val sessionId: UUID) : Signal
    data class Event(val sessionId: UUID, val event: ServerEvent) : Signal
    data class Lost(val sessionId: UUID, val error: NetworkException) : Signal
    data class Deadline(val sessionId: UUID, val deadlineId: UUID) : Signal
    data class Retry(val retryId: UUID) : Signal
}

private class MessagingFailureException(
    val failure: MessagingFailure,
) : Exception(failure.userMessage)

private fun Throwable.toMessagingFailure(): MessagingFailure = when (this) {
    is MessagingFailureException -> failure
    is PersistenceException -> MessagingFailure.Storage(this)
    else -> MessagingFailure.Network(NetworkException.wrap(this))
}
